### PR TITLE
Adds support for network latency metric.

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

Fixes: https://github.com/timescale/o11y-team-applications/issues/194

This PR adds a Prometheus metric that tracks the network latency from Promscale to the database.

## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [x] CHANGELOG entry for user-facing changes
- [x] Updated the relevant documentation

### CODE REVIEW

**Comment by niksajakovljevic** [comment]
@Harkishen-Singh Can you please explain in the description what do you mean by network latency.

File: docs/mixin/dashboards/promscale.json [hunk 1]
```diff
@@ -1,8 +1,8 @@
 {
   "__inputs": [
     {
-      "name": "DS_PROMETHEUS",
-      "label": "Prometheus",
+      "name": "DS_PROMSCALE_METRICS",
+      "label": "Promscale metrics",
```
Comment (paulfantom) [inline docs/mixin/dashboards/promscale.json:5]: Can we not change this?

File: docs/mixin/dashboards/promscale.json [hunk 2]
```diff
@@ -70,6 +73,10 @@
   "panels": [
     {
       "collapsed": false,
+      "datasource": {
+        "type": "prometheus",
+        "uid": "0yGYWPj7k"
+      },
```
Comment (paulfantom) [inline docs/mixin/dashboards/promscale.json:79]: I am not sure if static UID should be present here :thinking:

File: pkg/pgmodel/metrics/database/metrics.go [hunk 3]
```diff
@@ -30,25 +34,47 @@ func init() {
...
 type metricQueryWrap struct {
-	metric        prometheus.Collector
-	query         string
-	isHealthCheck bool
+	metric prometheus.Collector
+	query  string
 }
 
 var metrics = []metricQueryWrap{
-	{
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply pkg/pgmodel/metrics/database/metrics.go:40]: We should make it clear why we us another approach over existing one, or if needed we consider extending existing solution to support this use-case.

**Review by niksajakovljevic — COMMENTED** [review]
Let's discuss how to consolidate the code since we've introduced a snowflake.

File: pkg/pgmodel/metrics/database/metrics.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Harkishen-Singh) [reply pkg/pgmodel/metrics/database/metrics.go]: Now we use the existing approach, which also protects the /metrics from getting hammered.

[the remark this answers is not in the document]
Comment (Harkishen-Singh) [reply pkg/pgmodel/metrics/database/metrics.go]: Now we use the existing approach, which also protects the /metrics from getting hammered.

File: pkg/pgmodel/metrics/database/metrics.go [hunk 4]
```diff
@@ -30,25 +34,47 @@ func init() {
...
+			Subsystem:   "sql_database",
+			Name:        "network_latency_milliseconds",
+			Help:        "Network latency between Promscale and Database. A negative value indicates a failed health check.",
+			ConstLabels: map[string]string{"type": "promscale_sql"},
+		}, func() float64 {
+			// This function is called everytime Prometheus scrapes /metrics of Promscale.
+			// We use this to an advantage and do the health check as much as is expected by Prometheus.
+			val := 0
+			start := time.Now()
+			err := conn.QueryRow(context.Background(), "SELECT 1").Scan(&val)
```
[the remark this answers is not in the document]
Comment (Harkishen-Singh) [reply pkg/pgmodel/metrics/database/metrics.go]: Can you please suggest how to do a batch level SQL timeout? We do have context timeout, but there was a discussion to add a txn level statement_timeout. I tried looking into previously, but couldn't understand how to add a batch level statement_timeout.

[the remark this answers is not in the document]
Comment (Harkishen-Singh) [reply pkg/pgmodel/metrics/database/metrics.go]: Can you please suggest how to do a batch level SQL timeout? We do have context timeout, but there was a discussion to add a txn level statement_timeout. I tried looking into previously, but couldn't understand how to add a batch level statement_timeout.

File: docs/mixin/dashboards/promscale.json [hunk 5]
```diff
@@ -1227,658 +1282,787 @@
       "type": "row"
     },
     {
-      "collapsed": true,
+      "collapsed": false,
```
Comment (paulfantom) [inline docs/mixin/dashboards/promscale.json:1285]: Could you manually change this to `true` as it was before?

**Review by paulfantom — COMMENTED** [review]
I looked only at dashboard and it looks good to me apart from one minor issue mentioned in an inline comment.

File: pkg/pgmodel/metrics/database/database.go [hunk 6]
```diff
@@ -113,8 +113,36 @@ func (e *metricsEngineImpl) Run() error {
...
+		updateMetric(healthCheck.metric, 1)
+		if err != nil {
+			dbHealthErrors.Inc()
+			log.Error("msg", "health check failed", "err", err.Error())
+			// Important to set to -ve, otherwise if the connection is lost, latency will keep
+			// showing last latency of successful connection, if we choose to not update during
+			// an error.
+			networkLatency = -1
+		}
+		dbNetworkLatency.Set(float64(networkLatency))
```
Comment (antekresic) [inline pkg/pgmodel/metrics/database/database.go:145]: Wrap this into a function and call on line 120

**Review by antekresic — CHANGES_REQUESTED** [review]

**Review by antekresic — APPROVED** [review]

**Review by niksajakovljevic — APPROVED** [review]

**Comment by paulfantom** [comment]
Mixin CI check is failing due to https://github.com/timescale/promscale/issues/1446. You can safely ignore it for now.

**Pull request merged**